### PR TITLE
Timing & Sensitivity Screen

### PR DESCRIPTION
**Pull request by jsmorgan42**
Added dwell/hover time option & sensitivity options to the settings page.

Screen Shot 2020-03-31 at 10 03 48 AM
Screen Shot 2020-03-31 at 10 03 53 AM
Screen Shot 2020-03-31 at 10 01 04 AM
Screen Shot 2020-03-31 at 10 00 58 AM

### CODE REVIEW

File: Vocable/Features/Settings/TimingSensitivity/TimingSensitivityCollectionViewController.swift [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        collectionView.setCollectionViewLayout(createLayout(), animated: false)
+    }
+    
+    override func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        if collectionView.indexPathForGazedItem != indexPath {
+            collectionView.deselectItem(at: indexPath, animated: true)
+        }
+        
+        if AppConfig.isHeadTrackingEnabled {
+            //let alertViewController = GazeableAlertViewController.make { AppConfig.isHeadTrackingEnabled.toggle() }
```
Comment (thomasshealy) [inline Vocable/Features/Settings/TimingSensitivity/TimingSensitivityCollectionViewController.swift]: Commented code left in here.

File: Vocable/Features/Settings/TimingSensitivity/Views/SensitivityCollectionViewCell.swift [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+import UIKit
+import Combine
+
+class SensitivityCollectionViewCell: UICollectionViewCell {
+    
+    @IBOutlet var lowButton: GazeableSegmentedButton!
+    @IBOutlet var mediumButton: GazeableSegmentedButton!
+    @IBOutlet var highButton: GazeableSegmentedButton!
+
```
Comment (thomasshealy) [inline Vocable/Features/Settings/TimingSensitivity/Views/SensitivityCollectionViewCell.swift:17]: lowSensitivityButton, mediumSensitivityButton, and highSensitivityButton might be better names for these buttons.

File: Vocable/Features/Settings/TimingSensitivity/Views/SensitivityCollectionViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        
+        AppConfig.$sensitivity.sink { (sensitivity) in
+            UIView.performWithoutAnimation {
+                self.lowButton.isSelected = false
+                self.mediumButton.isSelected = false
+                self.highButton.isSelected = false
+
```
Comment (thomasshealy) [inline Vocable/Features/Settings/TimingSensitivity/Views/SensitivityCollectionViewCell.swift]: I think a switch might be better here.

**Review by thomasshealy — APPROVED** [review]
Just a few minor things, but LGTM.  I'd have Chris and/or Steve review as well before you merge

File: Vocable/AppConfig.swift [hunk 4]
```diff
@@ -24,4 +24,10 @@ struct AppConfig {
     static var isHeadTrackingSupported: Bool {
         return ARFaceTrackingConfiguration.isSupported
     }
+    
+    @PublishedDefault(key: "dwellTimeDuration", defaultValue: 1)
```
Comment (Clstroud) [inline Vocable/AppConfig.swift]: I've never felt like such a pedant, but `dwellTimeDuration` could be shortened to just `dwellDuration` 😅

File: Vocable/AppConfig.swift [hunk 5]
```diff
@@ -24,4 +24,10 @@ struct AppConfig {
...
+    static var selectionHoldDuration: TimeInterval
+    
+    @PublishedDefault(key: "sensitivitySetting", defaultValue: Sensitivity.medium)
+    static var sensitivity: Sensitivity
```
Comment (Clstroud) [inline Vocable/AppConfig.swift]: Might be good to make these preferences a bit clearer and pseudo namespace them e.g. `cursorSensitivity` for clarity. Same for the `Sensitivity` type. Future Jesse might appreciate slightly more specific naming 😉

File: Vocable/Common/Views/GazeableButton.swift [hunk 6]
```diff
@@ -177,3 +177,28 @@ class GazeableButton: UIButton {
...
+
+    override func gazeEnded(_ gaze: UIHeadGaze, with event: UIHeadGazeEvent?) {
+        if !isSelected {
+            gazeBeganDate = nil
+            isSelected = false
+        }
+        isHighlighted = false
+    }
+    
+}
```
Comment (Clstroud) [inline Vocable/Common/Views/GazeableButton.swift:204]: I expected this to end up being a lot worse, but this seems like a concise solution. Good stuff 👍

File: Vocable/Features/Settings/TimingSensitivity/Model/Sensitivity.swift [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    var range: ClosedRange<Double> {
+        switch self {
+        case .low:
+            return (2.0 ... 4.0)
+        case .medium:
+            return (3.0 ... 5.0)
+        case .high:
+            return (4.0 ... 6.5)
+        }
+    }
```
Comment (Clstroud) [inline Vocable/Features/Settings/TimingSensitivity/Model/Sensitivity.swift]: This definitely needs some sort of comment to explain what these numbers are doing. It was dense/confusing before, but now there's no context at all 😞

File: Vocable/Features/Settings/TimingSensitivity/Views/SensitivityCollectionViewCell.swift (hunk 3 above)
Comment (Clstroud) [reply Vocable/Features/Settings/TimingSensitivity/Views/SensitivityCollectionViewCell.swift]: `performWithoutAnimation` is probably unnecessary now that the button itself was refactored

File: Vocable/Features/Settings/TimingSensitivity/Views/SensitivityCollectionViewCell.swift [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        AppConfig.sensitivity = .low
+    }
+    
+    @IBAction func handleMediumSensitivity(_ sender: Any) {
+        AppConfig.sensitivity = .medium
+    }
+    
+    @IBAction func handleHighSensitivity(_ sender: Any) {
+        AppConfig.sensitivity = .high
+    }
```
Comment (Clstroud) [inline Vocable/Features/Settings/TimingSensitivity/Views/SensitivityCollectionViewCell.swift]: Can these + the button instances themselves be private?

File: Vocable/HeadGazeLib/UIHeadGazeViewController.swift [hunk 9]
```diff
@@ -40,6 +40,11 @@ class UIHeadGazeViewController: UIViewController, ARSessionDelegate, ARSCNViewDe
     private var computedScale: CGFloat = 0
     private var xAngleCorrectionAmount = 0.0
     private var yAngleCorrectionAmount = 0.0
+    
+    let distanceRange = (0.3 ... 0.6) // Distance from camera
```
Comment (Clstroud) [inline Vocable/HeadGazeLib/UIHeadGazeViewController.swift]: `distanceRange` probably didn't need to move? The comment will make more sense in the context of the function, or you could flesh out the comment to be more illustrative of what this range does compared to the other if you want to keep it here.

**Review by Clstroud — APPROVED** [review]
Handful of small comments, but this looks good to me 👍 Nice job

**Pull request merged**